_posts/2016-11-09-making-a-small-object.md

```diff
@@ -0,0 +1,144 @@
+---
+layout: post
+title: Making a Small Object
+excerpt: The thought process behind creating a simple object to help deal with complex ActiveRecord ORDER BY queries with a strong focus on the Single Responsibility Principal.
+published: true
+categories: SQL Ruby ActiveRecord
+---
+
+I get excited about small objects when I program. I often open up files that are hundreds of lines long with mixed responsibilities and lots of complexity.
+I recently had an opportunity to make a small object with a single responsibility. The object would be used to build the SQL fragment that gets passed as a parameter to ActiveRecord's `.order()` method.
+
+My table of data had several columns that the user could order by. Like in an Excel spreadsheet, they wanted the ordering of previous columns to persist when ordering by a new column.
+
+For example, if I first ordered by name, then ordered by location, the table should first be ordered A-Z by location, then A-Z by name.
+
+Something like this:
+
+```
+| Name    | Location | Hire Date |
++---------+----------+-----------+
+  Andy    | Alabama  | 10/10/2016
+  Anthony | Alabama  | 09/12/2014
+  Betty   | Alabama  | 05/22/2015
+```
+
+There were also subtle differences with how we handled NULL values that needed to be addressed.
+
+Normally, null values appear first when a column is sorted ascending. We wanted this reversed, appearing last when ascending and first when descending. The exception to the rule was the hire date column, where we wanted the nulls to show up first when ascending and last when descending.
+
+The PostgreSQL syntax to do this is simply to supply `NULLS FIRST` or `NULLS LAST` to the statement.
+
+Unfortunately, there were some empty strings in the location column that we wanted to treat the same as nulls.
+
+The syntax for this in PostgreSQL is pretty simple as well `NULL_IF(column_name, value)`. In our case, `NULL_IF(location, '')` would do the trick.
+
+If the user asked to order first by name ascending then by hire date desc then by location desc, I needed to somehow generate args that looks like this:
+
+```ruby
+args = [
+  "name ASC NULLS LAST",
+  "hire_date DESC NULLS LAST",
+  "NULL_IF(location, '') DESC NULLS FIRST"
+]
+
+query.order(*args)
+```
+
+All these little conditionals started making the controller code pretty difficult to read and understand, so I made a small object called `OrderBy`.
+
+It is used like this:
+
+```ruby
+OrderBy.new("location", nulls: :reversed, null_if: "").to_sql
+
+# => "NULLIF(employees.job_location, '') ASC NULLS LAST"
+```
+
+Its implementation is small and simple.
+
+```ruby
+class OrderBy
+  attr_accessor :direction
+
+  def initialize(column, direction: :asc, nulls: nil, null_if: nil)
+    @column    = column
+    @direction = direction
+    @nulls     = nulls
+    @null_if   = null_if
+  end
```

Comment:
In the same spirit as my comment above, it'd be great if you could write little notes about the various methods (as you feel appropriate), explaining why you designed each method that way.